src/node/worklet.rs

```diff
@@ -0,0 +1,321 @@
+                    );
+                }
+                output_channel_count
+            };
+
+            // Setup audio params, set initial values when supplied via parameter_data
+            let mut node_param_map = HashMap::new();
+            let mut processor_param_map = HashMap::new();
+            for mut param_descriptor in P::parameter_descriptors() {
+                let name = std::mem::take(&mut param_descriptor.name);
+                let (param, proc) = context.create_audio_param(param_descriptor, &registration);
+                if let Some(value) = parameter_data.get(&name) {
+                    param.set_value(*value as f32); // mismatch in spec f32 vs f64
+                }
+                node_param_map.insert(name.clone(), param);
+                processor_param_map.insert(name, proc);
+            }
+
+            let node = AudioWorkletNode {
+                registration,
+                channel_config: channel_config.into(),
+                number_of_inputs,
+                number_of_outputs,
+                audio_param_map: node_param_map,
+            };
+
+            // TODO make initialization of proc nicer
+            let mut proc = None;
+            let mut processor_options = Some(processor_options);
+            let render = AudioWorkletRenderer {
+                processor: Box::new(move |s, i, o, p| {
+                    if proc.is_none() {
+                        let opts = processor_options.take().unwrap();
+                        proc = Some(P::constructor(opts));
+                    }
+                    proc.as_mut().unwrap().process(s, i, o, p)
+                }),
+                audio_param_map: processor_param_map,
+                output_channel_count,
+            };
+
+            (node, Box::new(render))
+        })
+    }
+
+    pub fn parameters(&self) -> &HashMap<String, AudioParam> {
+        &self.audio_param_map
+    }
+}
+
+type ProcessCallback = dyn for<'a, 'b> FnMut(
+    &'b RenderScope,
+    &'b [&'b [&'a [f32]]],
+    &'b mut [&'b mut [&'a mut [f32]]],
+    AudioParamValues<'b>,
+) -> bool;
+
+struct AudioWorkletRenderer {
+    processor: Box<ProcessCallback>,
+    audio_param_map: HashMap<String, AudioParamId>,
+    output_channel_count: Vec<usize>,
+}
+
+// SAFETY:
+// The concrete AudioWorkletProcessor is instantiated inside the render thread and won't be sent
+// elsewhere. TODO how to express this in safe rust? Can we remove the Send bound from
+// AudioProcessor?
+unsafe impl Send for AudioWorkletRenderer {}
+
+impl AudioProcessor for AudioWorkletRenderer {
+    fn process(
+        &mut self,
+        inputs: &[AudioRenderQuantum],
+        outputs: &mut [AudioRenderQuantum],
+        params: crate::render::AudioParamValues<'_>,
+        scope: &RenderScope,
+    ) -> bool {
+        // Bear with me, to construct a &[&[&[f32]]] we first build a backing vector of all the
```

Comment:
To get rid of this allocation in safe rust is going to be tricky. But I will try a few more things